src/main/java/generic/Demo.java

```diff
@@ -0,0 +1,31 @@
+package generic;
+
+import circle.Cylinder;
+import shapes.Circle;
+import circle.Ring;
+import shapes.Shape;
+
+import java.util.Arrays;
+
+public class Demo {
+    public static void main(String[] args) {
+        Pair<String, Shape> objectsPair = new Pair<>("Simple Shape", new Shape("red", true));
+        String name = objectsPair.getFirst();
+        Shape shape = objectsPair.getSecond();
+        System.out.println(name);
+        System.out.println(shape);
+
+        Ring ringRed = new Ring(7, "red");
+        Ring ringBlue = new Ring(6, "blue");
+        System.out.println("Rings are equal: " + compareObjects(ringRed, ringBlue));
+
+        Cylinder cylinderRed = new Cylinder(6, 1, "red");
+        Cylinder cylinderBlue = new Cylinder(6, 1, "blue");
+        System.out.println("Cylinders are equal: " + compareObjects(cylinderRed, cylinderBlue));
+    }
+
+    public static <T extends Comparable<T>> boolean compareObjects(T object1, T object2) {
```

Comment:
It's a very good example to use inheritance with a `Comparable` to improve understanding of how generics work.
It's also a good example of how `super` is used :) 

Currently `compareObjects` returns a `boolean`, but what if we want to have a method that gives us `getMax` of two elements?
Like this: 
```
Cylinder cylinderRed = new Cylinder(6, 1, "red");
Cylinder cylinderBlue = new Cylinder(6, 1, "blue");
Cylinder bigger = getMax(cylinderRed, cylinderBlue);
System.out.println("This one is bigger: " + bigger);
```


```
public static <T extends Comparable<T>> T getMax(T object1, T object2) {
    boolean compareResult = true;
    return compareResult ? object1 : object2;
}
```
In such implementation `bigger` won't be of type `Cylinder` - it will be of type `Ring` - just because `T` should implement `Comparable<T>`, while `Cylinder` implements `Comparable<Ring>` because it extends `Ring`.
To make it work as we want, we need to define `getMax` to work with objects which implement `Comparable<any ancestor of T>` - so we come to 
```
public static <T extends Comparable<? super T>> T getMax(T object1, T object2) {}
```